blobstore.wit.md

```diff
@@ -0,0 +1,186 @@
+```
+
+// # wasi-cloud blobstore service definition
+//
+// A blobstore is service that can mange containers and objects.
+//
+// ## Object
+// An object is a named sequence of bytes. Objects can be read or written with a stream interface.
+// Once an object is written its contents cannot be modified.
+// Objects can be any size from 1 byte up to the limits imposed by the underlying store.
+// Object names are unique within their container.
+//
+// ## Container
+// A container is a named collection of objects. Container names can be any utf-8 string.
+// Within any instance of the blobstore interface, all containers share the same "flat" namespace.
+
+
+// wasi-cloud Blobstore service definition
+interface "wasi:blob/blobstore" {
+
+  use wasi:blob/types::{ Error, container-metadata, container-name, object-id }
+
+  // creates a new empty container
+  create-container: func(name: container-name) -> result<container,Error>
+
+  // get a list of all containers, as a stream.
+  list-containers: func() -> result<stream<container-name>, Error>
+
+  // retrieves a container by name
+  get-container: func(name: container-name) -> result<container, Error>
+
+  // deletes a container and all objects within it
+  delete-container: func(name: container-name) -> result<_, Error>
+
+  // returns true if the container exists
+  container-exists: func(name: container-name) -> result<bool, Error>
+
+  // copies (duplicates) an object, to the same or a different container.
+  // returns an error if the target container does not exist.
+  // overwrites destination object if it already existed.
+  copy-object: func(src: object-id, dest: object-id) -> result<_, Error>
+
+  // moves or renames an object, to the same or a different container
+  // returns an error if the destination container does not exist.
+  // overwrites destination object if it already existed.
+  move-object: func(src:object-id, dest: object-id) -> result<_, Error>
+}
+
+// a Container is a collection of objects
+resource container {
+
+  // returns container name
+  name: func() -> result<string, Error>
+
+  // returns container metadata
+  info: func() -> result<container-metadata, Error>
+
+  // begins reading an object
+  read-object: func(name: object-name) -> result<read-stream, Error>
+
+  // creates or replaces an object.
+  write-object: func(name: object-name) -> result<write-stream, Error>
+
+  // retrieves an object or portion of an object, as a resource.
+  // Start and end offsets are inclusive.
+  // Once a data-blob resource has been created, the underlying bytes are held by the blobstore service for the lifetime
+  // of the data-blob resource, even if the object they came from is later deleted.
+  get-data: func(name: object-name, start: u64, end: u64) -> result<data-blob, Error>
```

Comment:
This one is also drifting into edge case scenario. I don't think we can safely assume that this kind of operation is available on all blob store implementations. A number of them don't let you have random access, and so the implementation of this would be to grab the whole thing anyway, which could be a really really huge file and cause memory problems.

Reply:
there are two issues combined here, since the get-data is also the only method that returns a Resource handle.
- issue 1 is whether or not providers support byte range queries, and whether the api should be part of a core required functionality for all implementations. Do we need to do more research on how commonly that feature is supported? This kind of question will come up often in the design of the wasi-cloud apis and this seems like a good opportunity to understand how we make those decisions.

- issue 2 is around the use of Resource handle for a range of bytes, which has value regardless of whether or not the provider supports byte range requests. If we move or remove get-data due to lack of support for range requests, I'd still like to explore whether people think a resource handle for bytes is useful, and if yes, we can add back a method for get object that returns Resource.

Reply:
For issue #1, A provider that doesn't support this API could still provide an implementation by making this request ephemeral. The key for this is that it surfaces areas if the data will be larger than available memory.

Reply:
I did a quick survey to see what kinds of blob-like stores support byte ranges. It is supported by aws S3, most common file systems, Openstack Object store, and Postgres. Range is also a standard http header so a lot of http rest apis support it, and aws recommends asking for byte ranges as "best practice".

It's useful for
- when you know you need to process chunks of a file (although if we use streaming apis that's less often needed)
- if a download to local cache or storage was interrupted and you want to continue where you left off
- if the file contains fixed-sized records and you want to get a range of records
- if you want to load the header (first n bytes) to confirm file type, check if it's compressed or encrypted, etc. 
- if you are doing parallel download from multiple servers, as in bittorrent
- if the file contains a table of offsets (such as a directory portion of a zip file) so you could fetch the directory and then fetch the segment you want. 

It's not useful if data is stored as base64, utf8, json, etc. However in those cases the application isn't likely to use the byte range query anyway.

Reply:
Did you find any big implementers that don't support byte ranges?

Reply:
@danbugs I didn't find a counter-example, but since we're trying to make an api that will be used broadly and for future use cases that we haven't anticipated, I do think the api should have some way to support implementations that don't support ranges efficiently. Would keeping the range api as a distinct sub-interface accomplish that?